$OpenBSD: patch-poppler_Link_cc,v 1.1 2020/03/30 18:26:06 kili Exp $

Upstream commit 81a86064c14a7fc25047b6040d65464e732cf501 (Adam
Reichold):

Fix vague linkage of Link* class vtables

Due to falling back to the implicitly inline destructors, some of
the Link* classes had all their overridden methods defined inline
with made the linkage of their vtables vague.

This change moves their destructors into a defined translation unit
thereby anchoring their vtables in the libpoppler DSO which fixes
issues using dynamic_cast when builing Poppler using Clang.

Index: poppler/Link.cc
--- poppler/Link.cc.orig
+++ poppler/Link.cc
@@ -444,6 +444,8 @@ LinkGoTo::LinkGoTo(const Object *destObj) {
   }
 }
 
+LinkGoTo::~LinkGoTo() = default;
+
 //------------------------------------------------------------------------
 // LinkGoToR
 //------------------------------------------------------------------------
@@ -474,6 +476,8 @@ LinkGoToR::LinkGoToR(Object *fileSpecObj, Object *dest
   }
 }
 
+LinkGoToR::~LinkGoToR() = default;
+
 //------------------------------------------------------------------------
 // LinkLaunch
 //------------------------------------------------------------------------
@@ -552,6 +556,8 @@ LinkURI::LinkURI(const Object *uriObj, const GooString
   }
 }
 
+LinkURI::~LinkURI() = default;
+
 //------------------------------------------------------------------------
 // LinkNamed
 //------------------------------------------------------------------------
@@ -564,6 +570,8 @@ LinkNamed::LinkNamed(const Object *nameObj) {
   }
 }
 
+LinkNamed::~LinkNamed() = default;
+
 //------------------------------------------------------------------------
 // LinkMovie
 //------------------------------------------------------------------------
@@ -607,6 +615,8 @@ LinkMovie::LinkMovie(const Object *obj) {
   }
 }
 
+LinkMovie::~LinkMovie() = default;
+
 //------------------------------------------------------------------------
 // LinkSound
 //------------------------------------------------------------------------
@@ -645,6 +655,8 @@ LinkSound::LinkSound(const Object *soundObj) {
   }
 }
 
+LinkSound::~LinkSound() = default;
+
 //------------------------------------------------------------------------
 // LinkRendition
 //------------------------------------------------------------------------
@@ -738,6 +750,8 @@ LinkJavaScript::LinkJavaScript(Object *jsObj) {
   }
 }
 
+LinkJavaScript::~LinkJavaScript() = default;
+
 Object LinkJavaScript::createObject(XRef *xref, const GooString &js)
 {
   Dict *linkDict = new Dict(xref);
@@ -793,6 +807,8 @@ LinkOCGState::LinkOCGState(const Object *obj)
   preserveRB = obj->dictLookup("PreserveRB").getBoolWithDefaultValue(true);
 }
 
+LinkOCGState::~LinkOCGState() = default;
+
 //------------------------------------------------------------------------
 // LinkHide
 //------------------------------------------------------------------------
@@ -814,6 +830,8 @@ LinkHide::LinkHide(const Object *hideObj) {
   }
 }
 
+LinkHide::~LinkHide() = default;
+
 //------------------------------------------------------------------------
 // LinkUnknown
 //------------------------------------------------------------------------
@@ -821,6 +839,8 @@ LinkHide::LinkHide(const Object *hideObj) {
 LinkUnknown::LinkUnknown(const char *actionA) {
   action = std::string(actionA ? actionA : "");
 }
+
+LinkUnknown::~LinkUnknown() = default;
 
 //------------------------------------------------------------------------
 // Links
